[TEST] 도메인 유닛 테스트 작성

jd99iam opened the pull request:
## #222 

### 💡 PR 타입(하나 이상의 PR 타입을 선택해주세요)
* [ ] 기능 추가
* [ ] 기능 삭제
* [ ] 버그 수정
* [x] 테스트코드 작성
* [ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트

### 📲 반영 브랜치
test/domainUnitTest -> develop

### ❔ 변경 사항
일부 도메인에 대한 유닛 테스트를 작성했습니다 

### 🔍 테스트 결과
- Repo 테스트
- Service 테스트
- Controler 테스트

### 📰 참고 Reference
 
### ⚒️ 기타 사항
도메인 테스트 이런 식으로 작성하는 것에 대해 괜찮은지? 의견도 여쭙고 한번에 전부 작성하면 너무 많아져서 리뷰하기 힘드실 것 같아서 여기까지 하고 pr 한번 올리겠습니다.

jd99iam committed:
refactor: 팔로잉, 팔로워 수 얻는 메서드 작명 변경

getter와 혼동되는 이름이므로 재작명
- getFollowCount -> countFollow
- getFollowingCount -> countFollowing

related to: #222

jd99iam committed:
test(BadgeConstantTest): 뱃지 얻는 테스트 추가

작성한 풀이 갯수에 알맞게 뱃지를 얻을 수 있는지 테스트
- 각 풀이 작성 개수에 맞게 각 배지 enum을 반환하는지 테스트

related to: #222

jd99iam committed:
test(BadgeConstantTest): 뱃지 못 얻는 테스트 추가

지정한 풀이 개수에 해당하지 않으면 null을 반환하는지 테스트

related to: #222

jd99iam committed:
test(CommentTest): 댓글 수정 테스트 추가

댓글이 정상적으로 수정되는지 테스트
- 댓글의 내용이 정상수정되는지 테스트
- 댓글 상태가 FIXED로 수정되는지 테스트

related to: #222

jd99iam committed:
fix(Comment): 댓글 수정 로직 추가

댓글 수정 시 수정하려는 내용 isBlank로 검사

related to: #222

jd99iam committed:
fix(Comment): 댓글 수정 로직 수정

isBlank로 null도 검사 가능해서 null은 조건에서 제외

related to: #222

jd99iam committed:
fix(Comment): 댓글 수정 로직 수정

null이랑 isBlank 둘다 사용하도록 변경
- null이면 null.isBlank가 불가능
- null을 먼저 검사하도록 하기 위해 or의 앞부분에 위치
- 긍정 조건에 return하고 아닐 경우 수정하도록 함

related to: #222

jd99iam committed:
test(CommentTest): 댓글 수정 테스트 추가

댓글 수정 도메인 단위 테스트 추가
- 댓글 내용이 null 일 경우 수정할 수 없음
- 댓글 내용이 blank일 경우 수정할 수 없음

related to: #222

jd99iam committed:
test(CommentTest): 댓글 삭제 테스트 추가

댓글 삭제 도메인 단위 테스트 추가
- 댓글 삭제 시 내용 `삭제된 댓글입니다`로 변경
- 댓글 상태 `DELETED`로 변경

related to: #222

jd99iam committed:
refactor: 라인정리

related to: #222

jd99iam committed:
fix(Review): 리뷰 수정 로직 변경

리뷰 수정시 null검사, isBlank검사 하도록 수정

related to: #222

jd99iam committed:
test(ReviewTest): 리뷰 테스트 추가

댓글 단위 테스트랑 매우 유사해서 한번에 추가
- 리뷰 수정
- 리뷰 수정 불가
- 리뷰 삭제

related to: #222

jd99iam committed:
test(AlgorithmConstantTest): view->AlgorithmConstant 변환 테스트

변환 메서드 from 테스트
- CsvSource 사용

related to: #222

jd99iam committed:
test(AlgorithmConstantTest): view->AlgorithmConstant 변환 실패 테스트

변환 메서드 from 테스트
- 잘못된 뷰에 대해 `AlgorithmNotFoundException` 예외를 던짐

related to: #222

jd99iam committed:
test(DataStructureConstantTest): 자료구조 enum 테스트 추가

자료구조 테스트 추가
- 뷰 -> 상수 테스트 추가
- 잘못된 뷰에 대해 예외 발생 테스트 추가

related to: #220

jd99iam committed:
chore: 충돌 해결

develop 브랜치 pull 해서 변경사항 충돌 해결

related to: #222

jd99iam committed:
chore: 충돌 해결

develop 브랜치 pull 해서 변경사항 충돌 해결

related to: #222

jd99iam committed:
test: LanguageConstantTest 추가

뷰에 알맞은 언어를 얻을수 있는지 테스트 추가

related to: #222

jd99iam committed:
test: LanguageConstantTest 추가

부적절한 뷰에 대해 예외가 발생하는 지 테스트 추가

related to: #222

jd99iam committed:
refacotr: Solution 도메인 로직 수정

update 할 때 검사 로직 수정
- isBlank와 null 모두 검사하는 경우 null이 앞에 오게 or 조건을 활용해 검사
- null인 경우 isBlank를 호출하면 NPE가 터지기 때문
- null만 검사하는 경우는 `!=` 로 부정인지 검사

related to: #222

jd99iam committed:
refacotr: Solution 도메인 로직 수정

update 할 때 검사 로직 수정
- isBlank와 null 모두 검사하는 경우 null이 앞에 오게 or 조건을 활용해 검사
- null인 경우 isBlank를 호출하면 NPE가 터지기 때문
- null만 검사하는 경우는 `!=` 로 부정인지 검사

related to: #222

jd99iam committed:
test(Solution): 풀이 수정 정상 테스트 추가

풀이가 정상적으로 수정되는지 테스트

related to: #222

jd99iam committed:
feat(SolutionFixture): 풀이 픽스처에 알고리즘, 자료구조 지성 생성 메서드 추가

수정 여부를 확인하기 위해 픽스쳐에 알고리즘, 자료구조 지정해서 생성하는 메서드 추가

related to: #222

jd99iam committed:
test(SolutionTest): 수정하려는 제목이 null이면 제목이 수정되지 않는지 테스트

related to: #222

jd99iam committed:
test(SolutionTest): 수정하려는 제목이 isBlank이면 제목이 수정되지 않는지 테스트

related to: #222

jd99iam committed:
test(SolutionTest): 풀이 테스트 수정

update 실패 test 수정
- 값이 null이면 수정 안되는 것 필드 묶어서 한번에 하도록 수정
- 값이 isBlank이면 수정 안되는 것필드 묶어서 한번에 하도록 수정

related to: #222

jd99iam committed:
refactor(Solution): 라인정리

related to: #222

rlfrkdms1 commented on backend/src/test/java/com/prograngers/backend/entity/badge/BadgeConstantTest.java:
```diff
@@ -0,0 +1,28 @@
+                () -> assertThat(BadgeConstant.getBadge(100L)).isEqualTo(꽃),
+                () -> assertThat(BadgeConstant.getBadge(150L)).isEqualTo(화단)
+        );
+    }
```
테스트 코드가 변경에 많이 취약한 것 같아요 .. ! 
만약 풀이 개수가 변경된다면 더이상 작동하지 않을 것 같기 때문이에요 
다른 방식을 한번 고민해보는게 어떨까요 ?

jd99iam replied:
> 테스트 코드가 변경에 많이 취약한 것 같아요 .. ! 만약 풀이 개수가 변경된다면 더이상 작동하지 않을 것 같기 때문이에요 다른 방식을 한번 고민해보는게 어떨까요 ?

```
    @DisplayName("작성한 풀이 갯수에 알맞게 뱃지를 얻을 수 있다.")
    @ParameterizedTest
    @CsvSource({"새싹", "잎과줄기", "꽃봉오리", "꽃", "화단"})
    void getBadgeTest(BadgeConstant badgeConstant) {
        assertThat(BadgeConstant.getBadge(badgeConstant.getCount())).isEqualTo(badgeConstant);
    }
```

고민해 본 결과 이렇게 하면 `BadgeConstant`의 `count`가 변경되어도 테스트를 통과할  수 있습니다!
하지만 이렇게 작성하기 위해 `BadgeConstant`의 `getCount()`를 private -> public으로 변경했습니다.

rlfrkdms1 commented on backend/src/test/java/com/prograngers/backend/entity/comment/CommentTest.java:
```diff
@@ -44,4 +48,29 @@ void updateTest() {
+    @ValueSource(strings = {"", "    "})
+    void 수정하려는_내용이_blank일_경우_수정할_수_없다(String content) {
+        // when
+        comment.update(content);
```
```java
comment = 생성된_댓글.기본_정보_생성();
expect = 생성된_댓글.기본_정보_생성();

comment.update(null);

assertAll(
                () -> assertThat(comment.getContent()).isEqualTo(expect.getContent())
                ...
```

와 같이 작성하는게 어떨까요 ? 지금은 갑자기 comment의 content를 가져와 비교를 하니 CONTENT와 같다라고 되어있는데, 이렇게 되면 comment의 실제 content가 CONTENT의 내용과 같은지는 comment에 들어가봐야 알아서 depth가 늘어나는 것 같아요

jd99iam replied:
> ```java
> comment = 생성된_댓글.기본_정보_생성();
> expect = 생성된_댓글.기본_정보_생성();
> 
> comment.update(null);
> 
> assertAll(
>                 () -> assertThat(comment.getContent()).isEqualTo(expect.getContent())
>                 ...
> ```
> 
> 와 같이 작성하는게 어떨까요 ? 지금은 갑자기 comment의 content를 가져와 비교를 하니 CONTENT와 같다라고 되어있는데, 이렇게 되면 comment의 실제 content가 CONTENT의 내용과 같은지는 comment에 들어가봐야 알아서 depth가 늘어나는 것 같아요

좋은 지적 감사합니다! 
제가 코드를 작성할 때는 이런 문제를 못 느꼈는데, 타인이 봤을 때는 **해당 상수가 갖는 의미를 한번에 파악하기 어려울 것 같아요. 상수가 픽스처에 쓰이는지 들어가서 확인도 해야 하고요**.
말씀하신 대로 CommentFixture의 `수정된_댓글` ,`삭제된_댓글`을 활용하도록 테스트를 수정했습니다!

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/entity/solution/Solution.java:
```diff
@@ -80,33 +80,42 @@ public class Solution {
+        }
     }
 
     public void update(String title, AlgorithmConstant algorithm, DataStructureConstant dataStructure, int level,
```
이것들을 파라미터로 전부 받을 때는 그럴만한 이유가 있어야한다고 생각해요
클린 코드의 관점에서는 파라미터가 너무 많을 땐 하나의 class로 묶어서 전달하는게 좋다고 알고 있습니다. [관련링크](https://lordofkangs.tistory.com/196)도 첨부 할게요 !
그래서 update의 파라미터로 solution 객체 자체를 받는게 어떨까 싶어요 ! 어떻게 생각하실까요 ?!
변경하신다면 제 코드를 참고하시면 좋을 것 같습니다.

jd99iam replied:
> 이것들을 파라미터로 전부 받을 때는 그럴만한 이유가 있어야한다고 생각해요 클린 코드의 관점에서는 파라미터가 너무 많을 땐 하나의 class로 묶어서 전달하는게 좋다고 알고 있습니다. [관련링크](https://lordofkangs.tistory.com/196)도 첨부 할게요 ! 그래서 update의 파라미터로 solution 객체 자체를 받는게 어떨까 싶어요 ! 어떻게 생각하실까요 ?! 변경하신다면 제 코드를 참고하시면 좋을 것 같습니다.

좋은 내용 공유 감사합니다! 클린 코드 책을 사야겠어용
```
    public void update(UpdateSolutionRequest request) {
        updateTitle(request.getTitle());
        updateAlgorithm(request.getAlgorithm());
        updateDataStructure(request.getDataStructure());
        updateLevel(request.getLevel());
        updateCode(request.getCode());
        updateDescription(request.getDescription());
    }
```
update는 다음과 같이 수정했습니다. 왜냐하면 굳이 수정 dto를 entity로 변환할 필요가 없을 것 같아서요!

rlfrkdms1 replied:
음 .. 하지만 이렇게 변경한다면 기존에 Controller, Service까지던 DTO의 사용범위가 Domain까지 넓어지는 것 같아요.

jd99iam replied:
> 음 .. 하지만 이렇게 변경한다면 기존에 Controller, Service까지던 DTO의 사용범위가 Domain까지 넓어지는 것 같아요.

`solution.update(request.toSolution());`
서비스에서 엔티티로 변환해서 Solution에 넘겨주도록 수정했습니다!
이 방법이 클린 코드를 지키면서 저희끼리 정한 dto의 사용 범위도 지키는 방법인 것 같아요.
```
    public Solution toSolution() {
        return Solution.builder()
                .title(title)
                .algorithm(algorithm)
                .dataStructure(dataStructure)
                .code(code)
                .description(description)
                .level(level)
                .build();
    }
``` 
그런데 이 때 전달되는 Solution 객체는 단순히 풀이 수정을 위한 정보만 담고 있어서 어색한 느낌이 드는 것 같아요. 마치 엔티티를 dto 처럼 쓴 것 같아서요.. 저만 그렇게 느껴질 수도 있는데, 더 좋은 방법이 없을까요?
서비스단에서 풀이를 수정해버리자니 update~~ 메서드가 전부 private이라 그것도 불가능하네요

rlfrkdms1 reviewed (commented):
고생하셨어요 도메인 테스트 작성해보셨는데, 제가 남긴 코멘트는 클래스 하나에만 남겼지만 반복적으로 모든 클래스에서 보이는 문제점이에요 ! 이 점 유의해서 봐주시면 감사하겠습니다.

jd99iam committed:
refactor(BadgeConstant): BadgeConstant getCount 메서드를 public으로 변경

테스트에서 사용하기 위해 getCount의 접근지정자를 public으로 변경

related to: #222

jd99iam committed:
refactor(BadgeConstantTest): 변경에 취약한 테스트 수정

BadgeConstantTest 수정
- 뱃지에서 getCount를 통해 값을 가져오는 방식으로 변경
- 뱃지의 값이 변경되어도 테스트는 통과함

related to: #222

jd99iam committed:
refactor(CommentFixture): 댓글 픽스처 댓글 별 내용 수정

댓글 종류 별로 댓글 내용도 다르게 수정

related to: #222

jd99iam committed:
refactor(Comment): 삭제된 댓글 상수 public으로 접근지
정자 변경

CommentFixture에서 사용하기 위해 public으로 변경
- CommentFixture에서 Comment의 상수를 사용해야 Comment에서 사용하는 값과 같다는 것이 보장이 된다고 생각해서 Comment의 상수를 접근할 수 있게 하기 위해 접근지정자를 public으로 변경함

related to: #222

jd99iam committed:
refactor(CommentFixture): 댓글 픽스처 삭제된 댓글 내용 변경

삭제된 댓글은 내용이 Comment.DELETED_CONTENT여야 하므로 댓글 내용을 Comment의 상수를 가져와 변경

related to: #222

jd99iam committed:
refactor(CommentTest): 댓글 도메인 테스트 리팩터링

기존 테스트는 가독성이 떨어져서 수정
- 기존에는 CommentFixture의 댓글 내용을 상수로 CommentTest로 가져옴
- 수정한 댓글 내용, 삭제한 댓글 내용도 CommentTest에서 상수로 지정하고 사용함
- 지정해놓은 상수가 실제 픽스처에서 사용되는지 확인하기 위해 픽스처로 들어가야 하고 사용된다는 것을 유추해야 함
- 따라서 픽스처의 수정된 댓글, 삭제된 댓글을 활용해 테스트코드의 given 부분을 작성하고 이를 활용해서 then 부분에서 비교함

related to: #222

jd99iam committed:
refactor(Solution): 풀이 수정 메서드 변경

클린코딩을 위해 update 메서드 변경
- 매개변수 수를 1개로 줄임
- update의 용도가 수정 요청을 받아 풀이를 수정하는 것이므로 인자 형태를 UpdateSolutionRequest 1개 형태로 변경

related to: #222

jd99iam committed:
refactor(Solution): 풀이 수정 메서드 변경

클린코딩을 위해 update 메서드 변경
- 매개변수 수를 1개로 줄임
- update의 용도가 수정 요청을 받아 풀이를 수정하는 것이므로 인자 형태를 UpdateSolutionRequest 1개 형태로 변경
- 굳이 UpdateSolutionRequest -> Solution으로 변환하는 오버헤드를 줄임

related to: #222

jd99iam committed:
refactor(UpdateSolutionRequest): 로직 삭제

풀이를 update하는 로직이 dto에 있어서 삭제

related to: #222

jd99iam committed:
feat(UpdateSolutionRequest): dto를 엔티티로 변환하는 메서드 추가

Soltion.update()에서 인자로 dto를 받으면 dto의 사용범위가 domain까지 늘어나서 이를 방지하고자 엔티티로 변환해서 전달하기 위한 수정
- 변환 메서드 toSolution() 추가

related to: #222

jd99iam committed:
refactor(Solution): update 매개변수 형 변경

dto 사용 범위는 컨트롤러, 서비스 layer까지로 제한했기 때문에 엔티티를 매개변수로 받도록 변경
- update의 매개변수를 Solution으로 변경

related to: #222

jd99iam committed:
refactor(SolutionService): dto를 entity로 변환해서 전달

서비스계층에서 dto를 엔티티로 변환해서 도메인계층으로 전달함
- request.toSolution()을 해서 update에 전달하도록 수정

related to: #222

jd99iam commented:
> 고생하셨어요 도메인 테스트 작성해보셨는데, 제가 남긴 코멘트는 클래스 하나에만 남겼지만 반복적으로 모든 클래스에서 보이는 문제점이에요 ! 이 점 유의해서 봐주시면 감사하겠습니다.

넵 좋은 피드백 감사합니다! 일부만 작성하고 pr 올려서 가은님 의견을 듣길 잘한 것 같아요
코멘트 달아주신 파일은 수정해서 push 했고, 다른 도메인 테스트나 도메인도 비슷한 문제가 있는지 확인 후 수정사항 있으면 수정하고 push 하겠습니다! 감사합니다.

jd99iam committed:
refactor(SolutionTest): 풀이 수정 테스트 수정

비교할 값을 상수를 선언해서 사용하면 상수의 값의 출처와 실제 객체의 값이 그 값인지 알기 어려움
- 비교용 객체를 given에 생성하고 비교하는 방식으로 변경
- 불필요한 상수 선언 삭제

related to: #222

jd99iam committed:
refactor(MemberTest): 테스트 리팩터링

가독성 향상을 위한 리팩터링
- 기존에는 장지담의 nickname "jidam99"를 assert에서 비교함
- "jidam99"가 장지담의 닉네임인지 확인하려면 픽스처 장지담을 확인해야함
- 이런 과정 없이도 이해할 수 있도록 getNickname해서 비교하도록 변경
- 수정 전에는 Member에서 탈퇴한 사용자의 닉네임으로 반환하던 상수를 MemberTest에 필드로 선언하고 사용함
- 마찬가지로 이 값을 실제 Member에서 탈퇴한 사용자 대상으로 사용하는지 확인하려면 Member 객체를 들어가서 확인해야함
- 또한 Member의 상수값을 변경하면 실패한 테스트가 됨. 변경에 취약함
- 따라서 Member의 탈퇴한 사용자 상수를 public으로 변경하고 이 값을 사용해 비교하도록 변경함

related to: #222

jd99iam committed:
refactor(Member): 상수 접근지정자 변경

변경에 유연한 테스트를 작성하기 위해 탈퇴한 사용자 상수를 public으로 변경함

related to: #222

jd99iam committed:
refactor(Review): 상수 접근지정자 변경

변경에 유연한 테스트를 작성하기 위해 탈퇴한 사용자 상수를 public으로 변경함

related to: #222

jd99iam committed:
refactor(ReviewFixture): 리뷰 픽스처 내용 수정

리뷰 종류별로 content가 다르도록 픽스처 수정

related to: #222

jd99iam committed:
refactor(ReviewTest): 가독성을 위한 리팩터링

테스트 가독성을 위해 리팩터링
- 불필요한 상수 삭제
- given에서 비교를 위한 객체 생성해서 비교하도록 변경

related to: #222

jd99iam commented:
> > 고생하셨어요 도메인 테스트 작성해보셨는데, 제가 남긴 코멘트는 클래스 하나에만 남겼지만 반복적으로 모든 클래스에서 보이는 문제점이에요 ! 이 점 유의해서 봐주시면 감사하겠습니다.
> 
> 넵 좋은 피드백 감사합니다! 일부만 작성하고 pr 올려서 가은님 의견을 듣길 잘한 것 같아요 코멘트 달아주신 파일은 수정해서 push 했고, 다른 도메인 테스트나 도메인도 비슷한 문제가 있는지 확인 후 수정사항 있으면 수정하고 push 하겠습니다! 감사합니다.

작성한 테스트들에 대해 추가적인 수정 해서 push했습니다!
